homeassistant/components/acaia/sensor.py

```diff
@@ -0,0 +1,144 @@
+    AcaiaSensorEntityDescription(
+        key="weight",
+        translation_key="weight",
+        device_class=SensorDeviceClass.WEIGHT,
+        native_unit_of_measurement=UnitOfMass.GRAMS,
+        state_class=SensorStateClass.MEASUREMENT,
+        unit_fn=lambda data: (
+            UnitOfMass.OUNCES
+            if data.units == AcaiaUnitOfMass.OUNCES
+            else UnitOfMass.GRAMS
+        ),
+        value_fn=lambda scale: scale.weight,
+    ),
+)
+RESTORE_SENSORS: tuple[AcaiaSensorEntityDescription, ...] = (
+    AcaiaSensorEntityDescription(
+        key="battery",
+        translation_key="battery",
+        device_class=SensorDeviceClass.BATTERY,
+        native_unit_of_measurement=PERCENTAGE,
+        state_class=SensorStateClass.MEASUREMENT,
+        value_fn=lambda scale: (
+            scale.device_state.battery_level if scale.device_state else None
+        ),
+    ),
+)
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    entry: AcaiaConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up sensors."""
+
+    coordinator = entry.runtime_data
+    entities = [
+        AcaiaSensor(coordinator, entity_description) for entity_description in SENSORS
+    ]
+    entities.extend(
+        AcaiaRestoreSensor(coordinator, entity_description)
+        for entity_description in RESTORE_SENSORS
+    )
+    async_add_entities(entities)
+
+
+class AcaiaSensor(AcaiaEntity, SensorEntity):
+    """Representation of an Acaia sensor."""
+
+    entity_description: AcaiaSensorEntityDescription
+
+    def __init__(
+        self,
+        coordinator: AcaiaCoordinator,
+        entity_description: AcaiaSensorEntityDescription,
+    ) -> None:
+        """Initialize the sensor."""
+        super().__init__(coordinator, entity_description)
+
+    @property
+    def native_value(self) -> StateType:
+        """Return the state of the sensor."""
+        return self.entity_description.value_fn(self._scale)
+
+
+class AcaiaRestoreSensor(AcaiaSensor, RestoreSensor):
+    """Representation of an Acaia sensor with restore capabilities."""
+
+    def __init__(
+        self,
+        coordinator: AcaiaCoordinator,
+        entity_description: AcaiaSensorEntityDescription,
+    ) -> None:
+        """Initialize the sensor."""
+        super().__init__(coordinator, entity_description)
+        self._restored_data: SensorExtraStoredData | None = None
+
+    async def async_added_to_hass(self) -> None:
+        """Handle entity which will be added."""
+        await super().async_added_to_hass()
+        self._restored_data = await self.async_get_last_sensor_data()
+
+    @property
+    def native_value(self) -> StateType:
+        """Return the state of the sensor."""
+        if (value := self.entity_description.value_fn(self._scale)) is not None:
+            return value
+
+        if self._restored_data is None:
+            return None
+
+        return cast(StateType, self._restored_data.native_value)
```

Comment:
if the integration reloads, but we cannot connect to the scale we restore the value

Reply:
but yes, we do restore old data if we can't get a value, but I think that's pretty standard (this is like it is in shelly).Also,  I'd rather have a value here, even if it's an old one.

Reply:
I feel like we should not depend on the restored value forever. Say, I start HA, the value from last time gets restored. Fine. Now I attach the scale, and I see a new value. Fine. If I then remove the scale again, it will show the restored value again and not the new value. And if you then reload, it would show the new value.

So I would rather just say, use the shorthand attributes here and use _handle_coordinator_callback to just set it on load, and when you get new data, just keep using that data.

(Oh, now that I think of it, by restoring the restored value, that value will now be persisted for the reload, so after reloading it will show the same value as before)